File: tools/compose_scene.py
#!/usr/bin/env python3
"""Compose a map container's byte-exact tile banks into a viewable scene preview.

This is DISPLAY-ONLY. The byte-exact charblock/grid/metatile assets remain the
build source; this tool reads them and paints the composed 2D scene the map
actually renders, so a reviewer sees "four trees in a row" instead of a sheet of
tile fragments. It changes no build inputs and is never read back by the build.

Chain (all already extracted as clean-room assets):
  grid (128x128 cells)  -- value_low|value_high<<8 --> metatile index
  metatiles.tilemap     -- 492 metatiles, each 2x2 u16 tile entries
  tile entry            -- index = entry & 0x0fff, palette bank = entry >> 12
  index window          -- 0x000-0x1ff cb1, 0x200-0x3ff cb2, 0x400-0x5ff cb3,
                           0x600-0x7ff animation, 0x800-0xfff shared tileset
  charblock.banked.png  -- 8bpp where pixel & 0x0f is the real 4bpp index
  palette.NNN.png       -- 16 banks x 16 colours
"""
import argparse
import json
from pathlib import Path

from PIL import Image

ROOT = Path(__file__).resolve().parents[1]
TILE = 8
CELL_TILES = 2                      # metatile is 2x2 tiles
CELL = TILE * CELL_TILES           # 16 px per grid cell
GRID = 128                          # cells per side
WINDOW = 0x200                      # tiles per source window


def load_bank_indices(path):
    """Return (width_tiles, height_tiles, [tile_index_grid]) of 4bpp indices."""
    image = Image.open(path).convert("P")
    width, height = image.size
    pixels = image.load()
    tiles = []
    for ty in range(height // TILE):
        for tx in range(width // TILE):
            block = []
            for py in range(TILE):
                row = []
                for px in range(TILE):
                    row.append(pixels[tx * TILE + px, ty * TILE + py] & 0x0F)
                block.append(row)
            tiles.append(block)
    return tiles


def load_palette(path):
    image = Image.open(path).convert("RGB")
    width, height = image.size
    colors = []
    for y in range(height):
        for x in range(width):
            colors.append(image.getpixel((x, y)))
    return colors                    # 256 entries: bank*16 + index


def parse_metatiles(path):
    entries = [int(token, 16)
               for line in path.read_text().split("\n") if line.strip()
               for token in line.split()]
    return [entries[i:i + 4] for i in range(0, len(entries), 4)]


def cell_indices(grid_dir, metatile_count):
    # A cell's metatile index is value_low plus the low bits of value_high. Only
    # ceil(log2(metatile_count)) - 8 high bits are index; the remaining high bits
    # (e.g. 0x40/0x80 seen in the wild) are layer/priority attributes. Deriving
    # the width from the actual metatile count is essential: maps range from ~492
    # metatiles (1 high bit) to ~2250 (4 high bits), so a fixed mask mis-decodes
    # every cell whose index exceeds the assumed width.
    import math
    high_bits = max(0, math.ceil(math.log2(metatile_count)) - 8) if metatile_count > 1 else 0
    high_mask = (1 << high_bits) - 1
    low = Image.open(grid_dir / "value_low.png").convert("P").load()
    high = Image.open(grid_dir / "value_high.png").convert("P").load()
    return [[low[x, y] | ((high[x, y] & high_mask) << 8) for x in range(GRID)]
            for y in range(GRID)]


def standalone_banks():
    """The globally shared map tilesets, in resource order, grouped into runs
    of adjacent resources. A map loads one run at virtual tile base 0x800."""
    manifest = json.loads((ROOT / "assets/manifest.json").read_text())
    ids = []
    for series in manifest.get("series", []):
        if series.get("kind") == "golden-sun-standalone-tile-series":
            ids = [entry["id"] for entry in series["resources"]]
    runs, current = [], []
    for rid in ids:
        if current and int(rid, 16) != int(current[-1], 16) + 1:
            runs.append(current)
            current = []
        current.append(rid)
    if current:
        runs.append(current)
    return runs


def load_shared(ids):
    tiles = []
    for rid in ids:
        path = ROOT / f"assets/graphics/resource_{rid}/tiles.4bpp.png"
        if path.exists():
            tiles.extend(load_bank_indices(path))
    return tiles


def pick_shared(map_dir, gfx_dir, metatiles, cells, sources, palette):
    """Choose which shared tileset run this map loads.

    The map data carries no tileset id we have recovered, so select the run
    empirically: the correct tileset's tiles are continuous in colour with the
    map tiles they sit beside, while a wrong one paints blocks that clash. Tile
    STRUCTURE does not discriminate (wrong runs contain structured tiles too);
    colour continuity across the shared/normal boundary does.
    """
    best, best_score = None, None
    for run in standalone_banks():
        tiles = load_shared(run)
        if not tiles:
            continue
        total, count = 0.0, 0
        for cy in range(0, GRID, 2):          # sample every other cell: same
            for cx in range(0, GRID, 2):      # ranking, a quarter of the work
                mt = cells[cy][cx]
                if mt >= len(metatiles):
                    continue
                mine = cell_colour(metatiles[mt], sources, tiles, palette)
                if mine is None or not mine[1]:
                    continue
                for dx, dy in ((1, 0), (0, 1)):
                    nx, ny = cx + dx, cy + dy
                    if not (0 <= nx < GRID and 0 <= ny < GRID):
                        continue
                    nmt = cells[ny][nx]
                    if nmt >= len(metatiles):
                        continue
                    other = cell_colour(metatiles[nmt], sources, tiles, palette)
                    if other is None or other[1]:
                        continue
                    total += sum(abs(a - b) for a, b in zip(mine[0], other[0]))
                    count += 1
        if count:
            score = total / count
            if best_score is None or score < best_score:
                best, best_score = tiles, score
    return best


def cell_colour(quad, sources, shared, palette):
    """(mean RGB, uses_shared) for one metatile, or None if nothing painted."""
    acc = [0, 0, 0]
    seen = 0
    uses_shared = False
    for entry in quad:
        index = entry & 0x0FFF
        bank = entry >> 12
        window, local = index // WINDOW, index % WINDOW
        if window < 4 and window in sources and local < len(sources[window]):
            tile = sources[window][local]
        elif shared is not None and 0x800 <= index and (index - 0x800) < len(shared):
            tile = shared[index - 0x800]
            uses_shared = True
        else:
            continue
        for row in tile:
            for value in row:
                red, green, blue = palette[(bank * 16 + value) & 0xFF]
                acc[0] += red
                acc[1] += green
                acc[2] += blue
                seen += 1
    if not seen:
        return None
    return ([channel / seen for channel in acc], uses_shared)


def compose(map_dir, gfx_dir, shared_tiles=None):
    window_files = {
        0: "charblock1.banked.png",
        1: "charblock2.banked.png",
        2: "charblock3.banked.png",
        3: "animation_source.banked.png",
    }
    sources = {w: load_bank_indices(gfx_dir / name)
               for w, name in window_files.items()
               if (gfx_dir / name).exists()}
    palette_path = next(gfx_dir.glob("palette.*.png"))
    palette = load_palette(palette_path)
    metatiles = parse_metatiles(map_dir / "components" / "metatiles.tilemap")
    cells = cell_indices(map_dir / "grid", len(metatiles))
    if shared_tiles is None:
        shared_tiles = pick_shared(map_dir, gfx_dir, metatiles, cells,
                                   sources, palette)

    canvas = Image.new("RGB", (GRID * CELL, GRID * CELL), (0, 0, 0))
    out = canvas.load()
    missing_shared = 0
    for cy in range(GRID):
        for cx in range(GRID):
            mt = cells[cy][cx]
            if mt >= len(metatiles):
                continue
            quad = metatiles[mt]
            for sub in range(4):
                entry = quad[sub]
                index = entry & 0x0FFF
                bank = entry >> 12
                window = index // WINDOW
                local = index % WINDOW
                if window < 4 and window in sources and local < len(sources[window]):
                    tile = sources[window][local]
                elif (shared_tiles is not None and index >= 0x800
                      and (index - 0x800) < len(shared_tiles)):
                    tile = shared_tiles[index - 0x800]
                else:
                    missing_shared += 1
                    continue
                ox = cx * CELL + (sub % CELL_TILES) * TILE
                oy = cy * CELL + (sub // CELL_TILES) * TILE
                for py in range(TILE):
                    for px in range(TILE):
                        color = palette[(bank * 16 + tile[py][px]) & 0xFF]
                        out[ox + px, oy + py] = color
    return canvas, missing_shared


def self_contained_maps():
    """Map containers that own their charblocks (renderable without the shared
    tileset / cross-container family linkage, which is not yet recovered)."""
    ids = []
    for grid in sorted((ROOT / "assets/maps").glob("resource_*/grid")):
        cid = grid.parent.name.split("_")[1]
        gfx = ROOT / f"assets/graphics/resource_{cid}"
        if (gfx / "charblock1.banked.png").exists():
            ids.append(cid)
    return ids


def render(cid, out_dir, crop=True):
    map_dir = ROOT / f"assets/maps/resource_{cid}"
    gfx_dir = ROOT / f"assets/graphics/resource_{cid}"
    canvas, missing = compose(map_dir, gfx_dir)
    if crop:
        from PIL import ImageChops
        bbox = ImageChops.difference(
            canvas, Image.new("RGB", canvas.size, (0, 0, 0))).getbbox()
        if bbox:
            canvas = canvas.crop(bbox)
    out_dir.mkdir(parents=True, exist_ok=True)
    output = out_dir / f"scene_{cid}.png"
    canvas.save(output)
    return output, canvas.size, missing


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("container", nargs="?", help="hex id, e.g. 152")
    parser.add_argument("-o", "--output", type=Path)
    parser.add_argument("--all", action="store_true",
                        help="render every self-contained map to out/scenes/")
    args = parser.parse_args()
    out_dir = ROOT / "out/scenes"
    if args.all:
        ids = self_contained_maps()
        print(f"rendering {len(ids)} self-contained map scenes to {out_dir}/")
        for cid in ids:
            output, size, missing = render(cid, out_dir)
            print(f"  {cid}: {size[0]}x{size[1]} (unresolved shared refs={missing})")
        return
    if not args.container:
        parser.error("give a container id or --all")
    output, size, missing = render(args.container, out_dir)
    if args.output:
        Image.open(output).save(args.output)
        output = args.output
    print(f"wrote {output} ({size[0]}x{size[1]}), "
          f"unresolved shared-tile refs={missing}")


if __name__ == "__main__":
    main()

File: tools/compose_scene.ts
#!/usr/bin/env bun
// Tool role: both; imported by tools/bg_tile_usage.ts, tools/tile_objects.ts; invoked by CONVENTIONS.md.
/**
 * 地図コンテナの正規タイル素材を、確認用の一枚絵として合成する。
 *
 * 出力画像は表示確認専用であり、構築入力には使わない。
 * ヘッダ、読込表、グリッド、メタタイル、文字ブロック、パレットを
 * 実機の表示順に重ね、反転指定も各タイルへ適用する。
 */
import {
  copyFileSync,
  existsSync,
  mkdirSync,
  readFileSync,
  writeFileSync,
} from "node:fs";
import { dirname, extname, join, resolve } from "node:path";
import { deflateSync } from "./zlib.ts";

import { chunk } from "./export_asset.ts";
import { resourceGraphicsDir } from "./asset_paths.ts";
import { prune_files } from "./generated_files.ts";
import { indexed_png, type Rgb } from "./import_asset.ts";
import { records_by_container } from "./map_load_table.ts";
import { decode_entry } from "./text_bg.ts";
import { build_bank as build_object_bank, unpack_tile } from "./tile_objects.ts";

export const ROOT = resolve(dirname(Bun.fileURLToPath(import.meta.url)), "..");
export const TILE = 8;
export const CELL_TILES = 2;
export const CELL = TILE * CELL_TILES;
export const GRID = 128;
export const WINDOW = 0x200;
export const LOAD_TABLE = join(ROOT, "assets/maps/map_load_table.json");
export const MANIFEST = join(ROOT, "assets/manifest.json");

export type Tile = number[][];
export type Color = Rgb;
export type Linkage = {
  map_index: number;
  container: string;
  palette: string;
  vram_charblock1: string;
  vram_charblock2: string;
  vram_charblock3: string;
  animation_source: string;
};

export interface SceneImage {
  width: number;
  height: number;
  pixels: Buffer;
}

type Catalog = [Record<string, string>, Record<string, string>];
type SceneDimensions = [number, number, number, number];
type WindowOrder = [number, number];
type RenderResult = [string, [number, number], number];

function read_json(path: string): any {
  return JSON.parse(readFileSync(path, "utf8"));
}

/** ヘッダに記録された画面寸法と必要セル数を返す。 */
export function scene_dimensions(map_dir: string): SceneDimensions {
  const document = read_json(join(map_dir, "components/header.json"));
  const parameters = document.parameters ?? [];
  if (parameters.length !== 12) {
    throw new Error("map header requires twelve parameter bytes");
  }
  const width = Number(parameters[2]) * TILE;
  const height = Number(parameters[3]) * TILE;
  if (!(width > 0 && width <= GRID * CELL) || !(height > 0 && height <= GRID * CELL)) {
    throw new Error("map header dimensions exceed the 128x128-cell grid");
  }
  const cells_wide = Math.floor((width + CELL - 1) / CELL);
  const cells_high = Math.floor((height + CELL - 1) / CELL);
  return [width, height, cells_wide, cells_high];
}

/** ヘッダ指定から二個の表示窓へ割り当てる文字ブロック順を返す。 */
export function tile_window_order(map_dir: string): WindowOrder {
  const document = read_json(join(map_dir, "components/header.json"));
  const parameters = document.parameters ?? [];
  if (parameters.length !== 12) {
    throw new Error("map header requires twelve parameter bytes");
  }
  const selector = Number(parameters[7]);
  if (selector !== 0 && selector !== 1) {
    throw new Error("map header tile-window selector must be zero or one");
  }
  return selector === 0 ? [2, 1] : [1, 2];
}

/** 4bpp索引画像を8画素単位のタイル列として読む。 */
export function load_bank_indices(path: string): Tile[] {
  const [width, height, pixels] = indexed_png(readFileSync(path));
  const tiles: Tile[] = [];
  for (let ty = 0; ty < Math.floor(height / TILE); ty += 1) {
    for (let tx = 0; tx < Math.floor(width / TILE); tx += 1) {
      const block: Tile = [];
      for (let py = 0; py < TILE; py += 1) {
        const row: number[] = [];
        for (let px = 0; px < TILE; px += 1) {
          row.push(pixels[(ty * TILE + py) * width + tx * TILE + px] & 0x0f);
        }
        block.push(row);
      }
      tiles.push(block);
    }
  }
  return tiles;
}

/** 素材目録を読込資源番号ごとのパレット表と文字表へ展開する。 */
export function graphics_catalog(): Catalog {
  const manifest = read_json(MANIFEST);
  const palettes: Record<string, string> = {};
  const banks: Record<string, string> = {};
  for (const series of manifest.series ?? []) {
    const kind = series.kind;
    if (kind === "golden-sun-map-charblock-series") {
      for (const family of series.families) {
        const base = Number.parseInt(family.id, 16);
        const directory = join(ROOT, resourceGraphicsDir(family.id));
        palettes[(base + 1).toString(16)] = join(directory, "palette.224.png");
        for (const [index, resource] of family.charblocks.entries()) {
          banks[(base + 2 + index).toString(16)] = resource.source
            ? join(ROOT, resource.source)
            : join(directory, `charblock${index + 1}.4bpp.png`);
        }
        if (family.animation_source !== undefined && family.animation_source !== null) {
          banks[(base + 5).toString(16)] = join(directory, "animation_source.4bpp.png");
        }
      }
    } else if (kind === "golden-sun-standalone-palette-series") {
      for (const palette of series.palettes) {
        const resource = String(palette.id).toLowerCase();
        palettes[resource] = join(ROOT, resourceGraphicsDir(resource), "palette.224.png");
      }
    } else if (kind === "golden-sun-standalone-tile-series") {
      for (const bank of series.resources) {
        const resource = String(bank.id).toLowerCase();
        banks[resource] = join(ROOT, resourceGraphicsDir(resource), "tiles.4bpp.png");
      }
    }
  }
  return [palettes, banks];
}

/** 中立タイル画像または物体計画から表示用タイル列を読む。 */
export function load_bank_source(source: string): Tile[] {
  if (extname(source) === ".json") {
    const [packed] = build_object_bank(source);
    const tiles: Tile[] = [];
    for (let start = 0; start < packed.length; start += 32) {
      const flat = unpack_tile(packed.subarray(start, start + 32));
      const tile: Tile = [];
      for (let row = 0; row < TILE; row += 1) {
        tile.push(flat.slice(row * TILE, (row + 1) * TILE));
      }
      tiles.push(tile);
    }
    return tiles;
  }
  return load_bank_indices(source);
}

/** パレット画像を画素順の256色表として読む。 */
export function load_palette(path: string): Color[] {
  const [, , pixels, palette] = indexed_png(readFileSync(path));
  return pixels.map((index) => palette[index]);
}

/** 四要素ずつ並んだメタタイル表を読む。 */
export function parse_metatiles(path: string): number[][] {
  const entries = readFileSync(path, "utf8")
    .split("\n")
    .filter((line) => line.trim().length > 0)
    .flatMap((line) => line.trim().split(/\s+/).map((token) => Number.parseInt(token, 16)));
  const metatiles: number[][] = [];
  for (let index = 0; index < entries.length; index += 4) {
    metatiles.push(entries.slice(index, index + 4));
  }
  return metatiles;
}

/** 128×128グリッドから下位12bitのメタタイル番号を復元する。 */
export function cell_indices(grid_dir: string, metatile_count: number): number[][] {
  const [, , low] = indexed_png(readFileSync(join(grid_dir, "value_low.png")));
  const [, , high] = indexed_png(readFileSync(join(grid_dir, "value_high.png")));
  const cells: number[][] = [];
  for (let y = 0; y < GRID; y += 1) {
    const row: number[] = [];
    for (let x = 0; x < GRID; x += 1) {
      row.push(low[y * GRID + x] | (high[y * GRID + x] & 0x0f) << 8);
    }
    cells.push(row);
  }
  if (cells.some((row) => row.some((value) => value >= metatile_count))) {
    throw new Error("map grid references a metatile outside its table");
  }
  return cells;
}

/** 一件の地図を反転指定とパレット指定を含めて合成する。 */
export function compose(map_dir: string, linkage: Linkage): [SceneImage, number] {
  const [palettes, banks] = graphics_catalog();
  const sources = new Map<number, Tile[]>();
  for (const [window, charblock] of tile_window_order(map_dir).entries()) {
    const resource = linkage[`vram_charblock${charblock}` as keyof Linkage];
    if (typeof resource === "string" && banks[resource.toLowerCase()] !== undefined) {
      sources.set(window, load_bank_source(banks[resource.toLowerCase()]));
    }
  }
  const palette_path = palettes[linkage.palette.toLowerCase()];
  const palette = load_palette(palette_path);
  const metatiles = parse_metatiles(join(map_dir, "components/metatiles.tilemap"));
  const cells = cell_indices(join(map_dir, "grid"), metatiles.length);
  const [width, height, cells_wide, cells_high] = scene_dimensions(map_dir);
  const pixels = Buffer.alloc(width * height * 3);
  let missing_tiles = 0;
  for (let cy = 0; cy < cells_high; cy += 1) {
    for (let cx = 0; cx < cells_wide; cx += 1) {
      const metatile = cells[cy][cx];
      if (metatile >= metatiles.length) {
        continue;
      }
      const quad = metatiles[metatile];
      for (let sub = 0; sub < 4; sub += 1) {
        const [index, bank, hflip, vflip] = decode_entry(quad[sub]);
        const window = Math.floor(index / WINDOW);
        const local = index % WINDOW;
        const source = sources.get(window);
        if (source === undefined || local >= source.length) {
          missing_tiles += 1;
          continue;
        }
        const tile = source[local];
        const ox = cx * CELL + sub % CELL_TILES * TILE;
        const oy = cy * CELL + Math.floor(sub / CELL_TILES) * TILE;
        for (let py = 0; py < TILE; py += 1) {
          for (let px = 0; px < TILE; px += 1) {
            if (ox + px >= width || oy + py >= height) {
              continue;
            }
            const sx = hflip ? TILE - 1 - px : px;
            const sy = vflip ? TILE - 1 - py : py;
            const color = palette[(bank * 16 + tile[sy][sx]) & 0xff];
            if (color === undefined) {
              throw new Error("map tile references a missing palette color");
            }
            pixels.set(color, ((oy + py) * width + ox + px) * 3);
          }
        }
      }
    }
  }
  return [{ width, height, pixels }, missing_tiles];
}

/** 黒以外の画素を囲む最小矩形へ切り詰める。 */
function crop_black(image: SceneImage): SceneImage {
  let left = image.width;
  let top = image.height;
  let right = -1;
  let bottom = -1;
  for (let y = 0; y < image.height; y += 1) {
    for (let x = 0; x < image.width; x += 1) {
      const offset = (y * image.width + x) * 3;
      if (image.pixels[offset] !== 0 || image.pixels[offset + 1] !== 0 || image.pixels[offset + 2] !== 0) {
        left = Math.min(left, x);
        top = Math.min(top, y);
        right = Math.max(right, x);
        bottom = Math.max(bottom, y);
      }
    }
  }
  if (right < left || bottom < top) {
    return image;
  }
  const width = right - left + 1;
  const height = bottom - top + 1;
  const pixels = Buffer.alloc(width * height * 3);
  for (let y = 0; y < height; y += 1) {
    const source = ((top + y) * image.width + left) * 3;
    image.pixels.copy(pixels, y * width * 3, source, source + width * 3);
  }
  return { width, height, pixels };
}

/** RGB画素を非交差PNGとして保存する。 */
function write_rgb_png(path: string, image: SceneImage): void {
  const header = Buffer.alloc(13);
  header.writeUInt32BE(image.width, 0);
  header.writeUInt32BE(image.height, 4);
  header.set([8, 2, 0, 0, 0], 8);
  const rows: Buffer[] = [];
  for (let y = 0; y < image.height; y += 1) {
    rows.push(Buffer.from([0]));
    rows.push(image.pixels.subarray(y * image.width * 3, (y + 1) * image.width * 3));
  }
  const data = Buffer.concat([
    Buffer.from("89504e470d0a1a0a", "hex"),
    chunk(Buffer.from("IHDR"), header),
    chunk(Buffer.from("IDAT"), deflateSync(Buffer.concat(rows), { level: 9 })),
    chunk(Buffer.from("IEND"), Buffer.alloc(0)),
  ]);
  writeFileSync(path, data);
}

/** 地図と必要な画像素材がすべて揃う読込表の行だけを返す。 */
export function renderable_records(): Linkage[] {
  const [palettes, banks] = graphics_catalog();
  const records: Linkage[] = [];
  const grouped = records_by_container(LOAD_TABLE);
  const order: string[] = [];
  const seen = new Set<string>();
  for (const row of read_json(LOAD_TABLE).records ?? []) {
    const cid = String(row.container).toLowerCase();
    if (!seen.has(cid)) {
      seen.add(cid);
      order.push(cid);
    }
  }
  for (const cid of order) {
    const rows = grouped[cid] ?? [];
    for (const row of rows) {
      const linkage = row as Linkage;
      const map_dir = join(ROOT, `assets/maps/resource_${cid}`);
      const order = existsSync(map_dir) ? tile_window_order(map_dir) : [];
      if (
        existsSync(map_dir) &&
        palettes[linkage.palette.toLowerCase()] !== undefined &&
        order.every((number) => banks[linkage[`vram_charblock${number}` as keyof Linkage]!.toString().toLowerCase()] !== undefined)
      ) {
        records.push(linkage);
      }
    }
  }
  return records;
}

/** コンテナ番号と任意の地図番号から読込表の一行を選ぶ。 */
export function select_record(cid: string, map_index?: number): Linkage {
  let rows = (records_by_container(LOAD_TABLE)[cid.toLowerCase()] ?? []) as Linkage[];
  if (map_index !== undefined) {
    rows = rows.filter((row) => Number(row.map_index) === map_index);
  }
  if (rows.length === 0) {
    throw new Error(`no map load record for container ${cid}`);
  }
  if (rows.length > 1 && map_index === undefined) {
    const indices = rows.map((row) => String(row.map_index)).join(", ");
    throw new Error(`container ${cid} has map variants ${indices}; give --map-index`);
  }
  return rows[0];
}

/** 一件の地図を出力し、出力先、寸法、未解決タイル数を返す。 */
export function render(
  cid: string,
  out_dir: string,
  crop = true,
  linkage?: Linkage,
  filename?: string,
): RenderResult {
  const map_dir = join(ROOT, `assets/maps/resource_${cid}`);
  const [composed, missing] = compose(map_dir, linkage ?? select_record(cid));
  const canvas = crop ? crop_black(composed) : composed;
  mkdirSync(out_dir, { recursive: true });
  const output = join(out_dir, filename ?? `scene_${cid}.png`);
  write_rgb_png(output, canvas);
  return [output, [canvas.width, canvas.height], missing];
}

function usage(): never {
  throw new Error("give a container id or --all");
}

function print_help(): void {
  console.log(
    "usage: compose_scene.ts [-h] [-o OUTPUT] [--map-index MAP_INDEX] [--all] [container]\n\n" +
    "compose a reconstructed map into a display-only scene preview\n\n" +
    "positional arguments:\n" +
    "  container             hex container id, e.g. 152\n\n" +
    "options:\n" +
    "  -h, --help            show this help message and exit\n" +
    "  -o, --output OUTPUT   copy the rendered scene to OUTPUT\n" +
    "  --map-index MAP_INDEX select a loader-table variant\n" +
    "  --all                 render every loader-linked map",
  );
}

function integer(value: string, option: string): number {
  if (!/^[+-]?\d+$/.test(value)) {
    throw new Error(`${option} requires an integer`);
  }
  return Number(value);
}

/** 表示用地図画像の生成命令を処理する。 */
export function main(arguments_ = Bun.argv.slice(2)): void {
  if (arguments_.includes("-h") || arguments_.includes("--help")) {
    print_help();
    return;
  }
  let container: string | undefined;
  let output_path: string | undefined;
  let map_index: number | undefined;
  let render_all = false;
  for (let index = 0; index < arguments_.length; index += 1) {
    const argument = arguments_[index];
    if (argument === "--all") {
      render_all = true;
    } else if (argument === "-o" || argument === "--output") {
      if (index + 1 >= arguments_.length) {
        throw new Error(`${argument} requires a value`);
      }
      output_path = arguments_[++index];
    } else if (argument.startsWith("--output=")) {
      output_path = argument.slice("--output=".length);
    } else if (argument === "--map-index") {
      if (index + 1 >= arguments_.length) {
        throw new Error("--map-index requires a value");
      }
      map_index = integer(arguments_[++index], "--map-index");
    } else if (argument.startsWith("--map-index=")) {
      map_index = integer(argument.slice("--map-index=".length), "--map-index");
    } else if (argument.startsWith("-")) {
      throw new Error(`unrecognized argument: ${argument}`);
    } else if (container === undefined) {
      container = argument;
    } else {
      throw new Error(`unrecognized argument: ${argument}`);
    }
  }

  const out_dir = join(ROOT, "out/scenes");
  if (render_all) {
    const records = renderable_records();
    const counts = new Map<string, number>();
    for (const row of records) {
      const cid = row.container.toLowerCase();
      counts.set(cid, (counts.get(cid) ?? 0) + 1);
    }
    const expected = new Set<string>();
    console.log(`rendering ${records.length} loader-linked map scenes to ${out_dir}/`);
    for (const row of records) {
      const cid = row.container.toLowerCase();
      const filename = counts.get(cid) === 1
        ? `scene_${cid}.png`
        : `scene_${cid}_map${Number(row.map_index).toString().padStart(3, "0")}.png`;
      expected.add(filename);
      const [, size, missing] = render(cid, out_dir, true, row, filename);
      console.log(`  ${cid} map=${row.map_index}: ${size[0]}x${size[1]} (unresolved tiles=${missing})`);
    }
    const removed = prune_files(out_dir, "scene_*.png", expected);
    if (removed.length > 0) {
      console.log(`pruned stale scenes=${removed.length}`);
    }
    return;
  }
  if (container === undefined) {
    usage();
  }
  const linkage = select_record(container, map_index);
  let [output, size, missing] = render(container, out_dir, true, linkage);
  if (output_path !== undefined) {
    copyFileSync(output, output_path);
    output = output_path;
  }
  console.log(`wrote ${output} (${size[0]}x${size[1]}), unresolved tiles=${missing}`);
}

if (import.meta.main) {
  main();
}
